### PR TITLE
ENH: make closed part of IntervalDtype

### LINKED ISSUE #19371: API: Add closed to IntervalDtype
This was brought up by @shoyer in a different issue, and it makes sense to me

Adding `closed` to `IntervalDtype` makes sense to me, as in my mind `closed` is inherently part of the dtype.  Since differing `closed` makes two `IntervalIndex` incompatible, I'd expect their dtypes to not be equal, but they currently are, which seems a little strange:
```python
In [2]: ii1 = pd.interval_range(0, 3, closed='left')

In [3]: ii2 = pd.interval_range(0, 3, closed='right')

In [4]: ii1.dtype == ii2.dtype
Out[4]: True
```

There's also a larger discussion as to if adding `closed` to `IntervalDtype` would allow us to remove `closed` as a parameter to the `IntervalIndex` constructor.  My preference would be to keep the `closed` parameter for user convenience, similar to how `CategoricalIndex` accepts `categories` and `ordered` parameters despite also accepting `CategoricalDtype` (though maybe this is just for legacy reasons).  Willing to be convinced otherwise though.

xref https://github.com/pandas-dev/pandas/issues/19263#issuecomment-359986104 (original comment)
xref #19370 (might be rendered moot)

cc @shoyer

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #19371
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

cc @jschendel

Still need to figure out what to do for backwards compat for unpickled IntervalDtype objects; ATM this just sets closed=None.

This causes IntervalIndex.append to cast to object dtype instead of raising if we try to append a non-matching closed.  I think thats the right thing to do, but we haven't fully discussed that xref #37774

### CODE REVIEW

File: pandas/core/dtypes/dtypes.py [hunk 1]
```diff
@@ -1008,27 +1008,33 @@ class IntervalDtype(PandasExtensionDtype):
...
     num = 103
     _metadata = ("subtype",)
-    _match = re.compile(r"(I|i)nterval\[(?P<subtype>.+)\]")
+    _match = re.compile(
+        r"(I|i)nterval\[(?P<subtype>[^,]+)(, (?P<closed>(right|left|both|neither)))?\]"
+    )
     _cache: Dict[str_type, PandasExtensionDtype] = {}
 
-    def __new__(cls, subtype=None):
+    def __new__(cls, subtype=None, closed=None):
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:1016]: can you type

File: pandas/core/dtypes/dtypes.py [hunk 2]
```diff
@@ -1008,27 +1008,33 @@ class IntervalDtype(PandasExtensionDtype):
...
         from pandas.core.dtypes.common import is_string_dtype, pandas_dtype
 
         if isinstance(subtype, IntervalDtype):
             return subtype
+            # TODO: what if closed is also passed?
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:1021]: prob should raise (unless this is for pickle compat)

**Review by jreback — CHANGES_REQUESTED** [review]
conceptually on board

File: pandas/core/dtypes/dtypes.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:1021]: will do

**Comment by jbrockmendel** [comment]
updated to fix doctests, raise on mismatched 'closed'.  still not wild about how we're handling old pickles, and ATM still allowing closed=None to be passed to the constructor

**Comment by jbrockmendel** [comment]
mothballing to clear the queue while i figure out the pickle thing locally.

**Pull request closed without merging**